packages/reactive-element/src/test/decorators/initializer_decorators_test.ts

```diff
@@ -0,0 +1,222 @@
+  });
+
+  teardown(() => {
+    container.parentElement!.removeChild(container);
+  });
+
+  test('can create initializer decorator with `decorateProperty`', async () => {
+    const setProp = (value: string) =>
+      decorateProperty(
+        (ctor: ReactiveElementConstructor, name: PropertyKey) => {
+          ctor.addInitializer((e: A) => {
+            (e as any)[name as string] = value;
+          });
+        }
+      );
+
+    class A extends ReactiveElement {
+      @setProp('foo')
+      foo?: string;
+    }
+    customElements.define(generateElementName(), A);
+    const el = new A();
+    container.appendChild(el);
+    assert.equal(el.foo, 'foo');
+  });
+
+  test('can create controller decorator that uses hostConnected/hostDisconnected', async () => {
+    const listeners: WeakMap<
+      ReactiveElement,
+      Array<{type: string; listener: (e: Event) => void}>
+    > = new WeakMap();
+    const listenWindow = <T>(type: string) => {
+      return decorateProperty(
+        (ctor: ReactiveElementConstructor, name: PropertyKey) => {
+          ctor.addInitializer((e: ReactiveElement) => {
+            const listener = (event: Event) =>
+              ((((ctor.prototype as unknown) as T)[
+                name as keyof T
+              ] as unknown) as Function).call(e, event);
+            let l = listeners.get(e);
+            if (l === undefined) {
+              listeners.set(e, (l = []));
+              e.addController({
+                hostConnected() {
+                  l!.forEach((info) => {
+                    window.addEventListener(info.type, info.listener);
+                  });
+                },
+                hostDisconnected() {
+                  l!.forEach((info) => {
+                    window.removeEventListener(info.type, info.listener);
+                  });
+                },
+              });
+            }
+            l.push({type, listener});
+          });
+        }
+      );
+    };
+
+    class B extends ReactiveElement {
+      @listenWindow<B>('nug')
+      eventHandler1(e: Event) {
+        this.event1 = e.type;
+      }
+      @listenWindow<B>('zug')
+      eventHandler2(e: Event) {
+        this.event2 = e.type;
+      }
+      event1?: string;
+      event2?: string;
+    }
+    customElements.define(generateElementName(), B);
+    const el = new B();
+    container.appendChild(el);
+    document.body.dispatchEvent(new Event('nug', {bubbles: true}));
+    document.body.dispatchEvent(new Event('zug', {bubbles: true}));
+    assert.equal(el.event1, 'nug');
+    assert.equal(el.event2, 'zug');
+    el.event1 = undefined;
+    el.event2 = undefined;
+    container.removeChild(el);
+    document.body.dispatchEvent(new Event('nug', {bubbles: true}));
+    document.body.dispatchEvent(new Event('zug', {bubbles: true}));
+    assert.isUndefined(el.event1);
+    assert.isUndefined(el.event2);
+  });
+
+  test('can create controller decorator that uses hostUpdate', async () => {
```

Comment:
Do we need a test for every controller callback? Once we verify that an init runs, we know it can call `addController()` and that `addController()` is tested to implement the ReactiveControllerHost protocol.

Reply:
Changed test names.

Reply:
Do we need the tests at all though? I think only tests for `addInitializer()` are really needed, since an initializer is a callback on the instance and we already have tests that cover the `addController()` instance method.